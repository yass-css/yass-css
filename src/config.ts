import { FileSystem } from './file-system'
export interface UserConfig {
  src?: string;
  rules?: {
    namespace?: string;
    separator?: string;
  }
  stylesheet?: {
    buildPath?: string;
    filename?: string;
    include?: {
      baseClasses?: boolean;
      tokenClasses?: boolean;
      pseudos?: boolean;
    }
  },
}

export interface Config {
  src?: string[];
  rules: {
    namespace: string;
    separator: string;
  }
  stylesheet: {
    buildPath: string;
    filename: string;
    include: {
      baseClasses: boolean;
      tokenClasses: boolean;
      pseudos: boolean;
    }
  },
}

export const getConfig = (userConfig: Partial<UserConfig>): Config => {
  const { src } = userConfig
  const { buildPath, filename } = userConfig.stylesheet ?? {}
  const ignore = buildPath && filename ? [`${buildPath}/${filename}`] : []

  const config: Config = {
    src: src ? FileSystem.readDirectory(src, { ignore }) : undefined,
    rules: {
      namespace: userConfig?.rules?.namespace || '',
      separator: userConfig?.rules?.separator || ':',
    },
    stylesheet: {
      buildPath:  userConfig?.stylesheet?.buildPath || 'styles/yass/',
      filename: userConfig?.stylesheet?.filename || 'yass.css',
      include: {
        baseClasses: userConfig?.stylesheet?.include?.baseClasses !== undefined ? userConfig?.stylesheet?.include.baseClasses : true,
        tokenClasses: userConfig?.stylesheet?.include?.tokenClasses !== undefined ? userConfig?.stylesheet?.include.tokenClasses : true,
        pseudos: userConfig?.stylesheet?.include?.pseudos !== undefined ? userConfig?.stylesheet?.include.pseudos : true,
      }
    },
  }

  return config
}
