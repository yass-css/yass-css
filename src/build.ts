import { RootElement, StyleSheet } from './ast' 
import { Config } from './config'
import { AtomicClassTransformer, BaseCSSTransformer, CustomPropertyTransformer } from './transformers'
import rules from './definitions/css/rules.json'
import { validateTokens, validateRules } from './validation'

import type { DesignToken } from "./types"

export const build = (tokens: DesignToken[], config: Config): string => {
    if(!tokens || tokens.length === 0){
    return ''
  }  

  const validTokens = validateTokens(tokens)
  const validRules = validateRules(rules)

  const stylesheet = new StyleSheet([
    // Add the `:root` first
    new RootElement(
      // Add CSS Custom Properties to :root
      CustomPropertyTransformer.transform(validTokens, config)
    ),
    ...BaseCSSTransformer.transform(validRules, config),
    // Add the atomic classes
    ...AtomicClassTransformer.transform(validTokens, config),
  ]).toJSON()

  const { css } = stylesheet

  return css
}
