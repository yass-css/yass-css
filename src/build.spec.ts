import { build } from './build'
import { getConfig } from './config'
import type { DesignToken } from './types'
import type { Config } from './config'


describe('build()', () => {
  it('generates classname from token.key if token.name is not present', () => {
    const tokens: DesignToken[] = [
      {
        key: 'color-red',
        value: 'rgb(255, 0, 0)',
        properties: ['color'],
      },
      {
        key: 'color-green',
        value: 'rgb(0, 255, 0)',
        properties: ['color'],
      }      
    ]

    const config: Config = getConfig({})

    const stylesheet = build(tokens, config)
    expect(stylesheet).toContain('--color-red: rgb(255, 0, 0);')
    expect(stylesheet).toContain('--color-green: rgb(0, 255, 0);')
  })

  it('prefers token.name over token.key to generate classname if token.name is present', () => {
    const tokens: DesignToken[] = [{
      key: 'color-red',
      value: 'rgb(255, 0, 0)',
      name: 'red',
      properties: ['color'],
    }]

    const config: Config = getConfig({})
    const stylesheet = build(tokens, config)

    expect(stylesheet).toContain('--color-red: rgb(255, 0, 0);') // Uses token.key for the custom property
    expect(stylesheet).toContain('.color\\:red { color: var(--color-red); }') // Uses token.name for the classname
  })

  it('does not generate :root element if no tokens are defined', () => {
    const tokens: DesignToken[] = []

    const config: Config = getConfig({})
    const stylesheet = build(tokens, config)

    expect(stylesheet).toBe('')
  })

  it('matches snapshot', () => {
    const tokens: DesignToken[] = [
      {
        key: 'color-red-light',
        value: 'rgb(255, 0, 0)',
        properties: ['color'],
      },
      {
        key: 'color-red',
        value: 'rgb(255, 0, 0)',
        properties: ['color'],
        themes: {
          'light': '{color-red-light}'
        }
      },
      {
        key: 'color-green',
        value: 'rgb(0, 255, 0)',
        name: 'green',
        properties: ['color'],
      },
      {
        key: 'block',
        value: 'block',
        properties: ['display'],
      },
      {
        key: 'color-blue',
        value: 'rgb(0, 0, 200)',
        themes: {
          'high-contrast': 'rgb(0, 0, 255)',
          'dark': 'rgb(0, 0, 100)',
          'light': '{color-blue}'
        },
        name: 'blue',
        properties: ['color'],
      },      
    ]

    const config: Config = getConfig({
      rules: {
        namespace: 'yass-',
        separator: '-'
      }
    })
    const stylesheet = build(tokens, config)

    expect(stylesheet).toMatchSnapshot()
  })
})
