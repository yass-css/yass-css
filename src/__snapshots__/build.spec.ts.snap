// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build() matches snapshot 1`] = `
":root {
  --yass-color-red-light: rgb(255, 0, 0);
  --yass-color-red: rgb(255, 0, 0);
  --yass-color-green: rgb(0, 255, 0);
  --yass-block: block;
  --yass-color-blue: rgb(0, 0, 200);
}
.yass-color-color-red-light { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:active:active { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:autofill:autofill { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:checked:checked { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:default:default { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:defined:defined { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:disabled:disabled { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:enabled:enabled { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:first:first { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:first-child:first-child { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:first-of-type:first-of-type { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:fullscreen:fullscreen { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:focus:focus { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:focus-visible:focus-visible { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:focus-within:focus-within { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:hover:hover { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:indeterminate:indeterminate { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:invalid:invalid { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:last-child:last-child { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:last-of-type:last-of-type { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:link:link { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:only-of-type:only-of-type { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:optional:optional { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:required:required { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:valid:valid { color: var(--yass-color-red-light); }
.yass-color-color-red-light\\:visited:visited { color: var(--yass-color-red-light); }
.yass-color-color-red { color: var(--yass-color-red); }
.yass-color-color-red\\:active:active { color: var(--yass-color-red); }
.yass-color-color-red\\:autofill:autofill { color: var(--yass-color-red); }
.yass-color-color-red\\:checked:checked { color: var(--yass-color-red); }
.yass-color-color-red\\:default:default { color: var(--yass-color-red); }
.yass-color-color-red\\:defined:defined { color: var(--yass-color-red); }
.yass-color-color-red\\:disabled:disabled { color: var(--yass-color-red); }
.yass-color-color-red\\:enabled:enabled { color: var(--yass-color-red); }
.yass-color-color-red\\:first:first { color: var(--yass-color-red); }
.yass-color-color-red\\:first-child:first-child { color: var(--yass-color-red); }
.yass-color-color-red\\:first-of-type:first-of-type { color: var(--yass-color-red); }
.yass-color-color-red\\:fullscreen:fullscreen { color: var(--yass-color-red); }
.yass-color-color-red\\:focus:focus { color: var(--yass-color-red); }
.yass-color-color-red\\:focus-visible:focus-visible { color: var(--yass-color-red); }
.yass-color-color-red\\:focus-within:focus-within { color: var(--yass-color-red); }
.yass-color-color-red\\:hover:hover { color: var(--yass-color-red); }
.yass-color-color-red\\:indeterminate:indeterminate { color: var(--yass-color-red); }
.yass-color-color-red\\:invalid:invalid { color: var(--yass-color-red); }
.yass-color-color-red\\:last-child:last-child { color: var(--yass-color-red); }
.yass-color-color-red\\:last-of-type:last-of-type { color: var(--yass-color-red); }
.yass-color-color-red\\:link:link { color: var(--yass-color-red); }
.yass-color-color-red\\:only-of-type:only-of-type { color: var(--yass-color-red); }
.yass-color-color-red\\:optional:optional { color: var(--yass-color-red); }
.yass-color-color-red\\:required:required { color: var(--yass-color-red); }
.yass-color-color-red\\:valid:valid { color: var(--yass-color-red); }
.yass-color-color-red\\:visited:visited { color: var(--yass-color-red); }
.yass-color-green { color: var(--yass-color-green); }
.yass-color-green\\:active:active { color: var(--yass-color-green); }
.yass-color-green\\:autofill:autofill { color: var(--yass-color-green); }
.yass-color-green\\:checked:checked { color: var(--yass-color-green); }
.yass-color-green\\:default:default { color: var(--yass-color-green); }
.yass-color-green\\:defined:defined { color: var(--yass-color-green); }
.yass-color-green\\:disabled:disabled { color: var(--yass-color-green); }
.yass-color-green\\:enabled:enabled { color: var(--yass-color-green); }
.yass-color-green\\:first:first { color: var(--yass-color-green); }
.yass-color-green\\:first-child:first-child { color: var(--yass-color-green); }
.yass-color-green\\:first-of-type:first-of-type { color: var(--yass-color-green); }
.yass-color-green\\:fullscreen:fullscreen { color: var(--yass-color-green); }
.yass-color-green\\:focus:focus { color: var(--yass-color-green); }
.yass-color-green\\:focus-visible:focus-visible { color: var(--yass-color-green); }
.yass-color-green\\:focus-within:focus-within { color: var(--yass-color-green); }
.yass-color-green\\:hover:hover { color: var(--yass-color-green); }
.yass-color-green\\:indeterminate:indeterminate { color: var(--yass-color-green); }
.yass-color-green\\:invalid:invalid { color: var(--yass-color-green); }
.yass-color-green\\:last-child:last-child { color: var(--yass-color-green); }
.yass-color-green\\:last-of-type:last-of-type { color: var(--yass-color-green); }
.yass-color-green\\:link:link { color: var(--yass-color-green); }
.yass-color-green\\:only-of-type:only-of-type { color: var(--yass-color-green); }
.yass-color-green\\:optional:optional { color: var(--yass-color-green); }
.yass-color-green\\:required:required { color: var(--yass-color-green); }
.yass-color-green\\:valid:valid { color: var(--yass-color-green); }
.yass-color-green\\:visited:visited { color: var(--yass-color-green); }
.yass-display-block { display: var(--yass-block); }
.yass-display-block\\:active:active { display: var(--yass-block); }
.yass-display-block\\:autofill:autofill { display: var(--yass-block); }
.yass-display-block\\:checked:checked { display: var(--yass-block); }
.yass-display-block\\:default:default { display: var(--yass-block); }
.yass-display-block\\:defined:defined { display: var(--yass-block); }
.yass-display-block\\:disabled:disabled { display: var(--yass-block); }
.yass-display-block\\:enabled:enabled { display: var(--yass-block); }
.yass-display-block\\:first:first { display: var(--yass-block); }
.yass-display-block\\:first-child:first-child { display: var(--yass-block); }
.yass-display-block\\:first-of-type:first-of-type { display: var(--yass-block); }
.yass-display-block\\:fullscreen:fullscreen { display: var(--yass-block); }
.yass-display-block\\:focus:focus { display: var(--yass-block); }
.yass-display-block\\:focus-visible:focus-visible { display: var(--yass-block); }
.yass-display-block\\:focus-within:focus-within { display: var(--yass-block); }
.yass-display-block\\:hover:hover { display: var(--yass-block); }
.yass-display-block\\:indeterminate:indeterminate { display: var(--yass-block); }
.yass-display-block\\:invalid:invalid { display: var(--yass-block); }
.yass-display-block\\:last-child:last-child { display: var(--yass-block); }
.yass-display-block\\:last-of-type:last-of-type { display: var(--yass-block); }
.yass-display-block\\:link:link { display: var(--yass-block); }
.yass-display-block\\:only-of-type:only-of-type { display: var(--yass-block); }
.yass-display-block\\:optional:optional { display: var(--yass-block); }
.yass-display-block\\:required:required { display: var(--yass-block); }
.yass-display-block\\:valid:valid { display: var(--yass-block); }
.yass-display-block\\:visited:visited { display: var(--yass-block); }
.yass-color-blue { color: var(--yass-color-blue); }
.yass-color-blue\\:active:active { color: var(--yass-color-blue); }
.yass-color-blue\\:autofill:autofill { color: var(--yass-color-blue); }
.yass-color-blue\\:checked:checked { color: var(--yass-color-blue); }
.yass-color-blue\\:default:default { color: var(--yass-color-blue); }
.yass-color-blue\\:defined:defined { color: var(--yass-color-blue); }
.yass-color-blue\\:disabled:disabled { color: var(--yass-color-blue); }
.yass-color-blue\\:enabled:enabled { color: var(--yass-color-blue); }
.yass-color-blue\\:first:first { color: var(--yass-color-blue); }
.yass-color-blue\\:first-child:first-child { color: var(--yass-color-blue); }
.yass-color-blue\\:first-of-type:first-of-type { color: var(--yass-color-blue); }
.yass-color-blue\\:fullscreen:fullscreen { color: var(--yass-color-blue); }
.yass-color-blue\\:focus:focus { color: var(--yass-color-blue); }
.yass-color-blue\\:focus-visible:focus-visible { color: var(--yass-color-blue); }
.yass-color-blue\\:focus-within:focus-within { color: var(--yass-color-blue); }
.yass-color-blue\\:hover:hover { color: var(--yass-color-blue); }
.yass-color-blue\\:indeterminate:indeterminate { color: var(--yass-color-blue); }
.yass-color-blue\\:invalid:invalid { color: var(--yass-color-blue); }
.yass-color-blue\\:last-child:last-child { color: var(--yass-color-blue); }
.yass-color-blue\\:last-of-type:last-of-type { color: var(--yass-color-blue); }
.yass-color-blue\\:link:link { color: var(--yass-color-blue); }
.yass-color-blue\\:only-of-type:only-of-type { color: var(--yass-color-blue); }
.yass-color-blue\\:optional:optional { color: var(--yass-color-blue); }
.yass-color-blue\\:required:required { color: var(--yass-color-blue); }
.yass-color-blue\\:valid:valid { color: var(--yass-color-blue); }
.yass-color-blue\\:visited:visited { color: var(--yass-color-blue); }"
`;
